src/Sarif/Visitors/VersionOneLogicalLocationKeyToFullyQualifiedNameMappingVisitor.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT        
+// license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Sarif.VersionOne;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+    /// <summary>
+    /// This class visits all locations in a SARIF v1 log file. On observing use of result.logicalLocationKey,
+    /// which is used to disambiguate logical locations in the result.logicalLocations dictionary that have a 
+    /// common fully qualified name (but which are different types), the visitor creates a mapping between
+    /// the logical location key and its associated fully qualified name. This allows the v2 transformation
+    /// in particular to more easily populate its logical location equivalents. Additionally, the visitor 
+    /// stores a mapping from fully qualified name to decorated name, if one exists. These data (the
+    /// decorated name and fully qualified name) moved from the location object to the logical location
+    /// object in v2.
+    /// </summary>
+    public class VersionOneLogicalLocationKeyToFullyQualifiedNameMappingVisitor : SarifRewritingVisitorVersionOne
+    {
+        public VersionOneLogicalLocationKeyToFullyQualifiedNameMappingVisitor()
+        {
+            FullyQualifiedNameToDecoratedNameMap = new Dictionary<string, string>();
+            LogicalLocationKeyToFullyQualifiedNameMap = new Dictionary<string, string>();
+        }
+        public IDictionary<string, string> FullyQualifiedNameToDecoratedNameMap { get; set; }
+
+        public IDictionary<string, string> LogicalLocationKeyToFullyQualifiedNameMap { get; set; }
+
+        public override LocationVersionOne VisitLocationVersionOne(LocationVersionOne node)
+        {
+            if (!string.IsNullOrEmpty(node.LogicalLocationKey) &&
+                !string.IsNullOrEmpty(node.FullyQualifiedLogicalName) &&
+                !node.FullyQualifiedLogicalName.Equals(node.LogicalLocationKey))
```

Comment:
Crud. This test occurs three times, but you can't DRY it out to `static bool NodeHasAmbiguousLogicalLocation` because the node is of three different types. Still yet another validation of the design improvement where we moved these properties into the `Location` object.

Reply:
Indeed. Almost left a comment for you on precisely this point (my inability to DRY due to the 'duck typing' of our previous design that inlined FQN and logical location keys. :)

---
In reply to: [244240968](https://github.com/Microsoft/sarif-sdk/pull/1185#discussion_r244240968) [](ancestors = 244240968)